java/arcs/core/entity/EntityBase.kt

```diff
@@ -147,6 +147,19 @@ open class EntityBase(
     /** Returns the [FieldType] for the given collection field, or null if it does not exist. */
     private fun getCollectionTypeOrNull(field: String) = schema.fields.collections[field]
 
+    /** Copies all singleton and collection values into another [EntityBase]. */
+    protected fun copyInto(that: EntityBase, overwrite: Boolean = false) {
+        if (overwrite) {
+            that.singletons.putAll(this.singletons)
+            that.collections.putAll(this.collections)
+            return
+        }
+        this.singletons.forEach { that.singletons.putIfAbsent(it.key, it.value) }
+        this.collections.forEach {
+            that.collections.merge(it.key, it.value) { t, u -> if (t.isEmpty()) u else t }
```

Comment:
2 worries:

1) This is really hard to read.
2) If empty Set is equivalent to field not yet set, it may be hard to distinguish between:
```
// Make a copy with empty set of values.
entity.copy(values=setOf())
```
and:
```
// Make a full copy.
entity.copy()
```

copy() should maybe have nulls as default values, and then copy over data from the source entitiy for all fields that were null as arguments to `copy()`, wdyt?